plugins/VersionUpgrade/VersionUpgrade44to45/VersionUpgrade44to45.py

```diff
@@ -16,6 +26,79 @@
+        # If you're starting from an earlier version, you can't have had the bug that produces too many hidden stacks (https://github.com/Ultimaker/Cura/issues/6731).
+        # If you're starting from a later version, the bug was already fixed.
+        data_storage_root = os.path.dirname(Resources.getDataStoragePath())
+        folders = set(os.listdir(data_storage_root))  # All version folders.
+        folders = set(filter(lambda p: re.fullmatch(r"\d+\.\d+", p), folders))  # Only folders with a correct version number as name.
+        folders.difference_update({os.path.basename(Resources.getDataStoragePath())})  # Remove current version from candidates (since the folder was just copied).
+        if folders:
+            latest_version = max(folders, key = Version)  # Sort them by semantic version numbering.
+            if latest_version == "4.4":
+                self.removeHiddenStacks()
+
+    def removeHiddenStacks(self) -> None:
+        """
+        If starting the upgrade from 4.4, this will remove any hidden printer
+        stacks from the configuration folder as well as all of the user profiles
+        and definition changes profiles.
+
+        This will ONLY run when upgrading from 4.4, not when e.g. upgrading from
+        4.3 to 4.6 (through 4.4). This is because it's to fix a bug
+        (https://github.com/Ultimaker/Cura/issues/6731) that occurred in 4.4
+        only, so only there will it have hidden stacks that need to be deleted.
+        If people upgrade from 4.3 they don't need to be deleted. If people
+        upgrade from 4.5 they have already been deleted previously or never got
+        the broken hidden stacks.
+        """
+        Logger.log("d", "Removing all hidden container stacks.")
+        hidden_global_stacks = set()  # Which global stacks have been found? We'll delete anything referred to by these. Set of stack IDs.
+        hidden_extruder_stacks = set()  # Which extruder stacks refer to the hidden global profiles?
+        hidden_instance_containers = set()  # Which instance containers are referred to by the hidden stacks?
+
+        # First find all of the hidden container stacks.
+        data_storage = Resources.getDataStoragePath()
+        for root, _, files in os.walk(data_storage):
+            for filename in fnmatch.filter(files, "*.global.cfg"):
+                parser = configparser.ConfigParser(interpolation = None)
+                parser.read(os.path.join(root, filename))
+                if "metadata" in parser and "hidden" in parser["metadata"] and parser["metadata"]["hidden"] == "True":
+                    stack_id = urllib.parse.unquote_plus(os.path.basename(filename).split(".")[0])
+                    hidden_global_stacks.add(stack_id)
+                    # The user container and definition changes container are specific to this stack. We need to delete those too.
+                    if "containers" in parser:
+                        if "0" in parser["containers"]:
```

Comment:
These are the stack entries by "layer", right? eg layer 6 might contain the extruder. Might need a comment here

Reply:
Yeah they are. I'll clarify which stack entry is which container type.